data_transfer/lib/vttsma.py

```diff
@@ -0,0 +1,90 @@
+# NOTE: some methods exist for convience and will be advanced
+# for the CVS, e.g. serial_by_device_uuid might to hit a live 
+# endpoint rather than pull from a CSV file.
+
+from data_transfer.config import config
+from data_transfer.services import dmpy
+from pathlib import Path
+from dataclasses import dataclass
+from typing import Optional
+from functools import lru_cache  # note: @cache exists in py >= 3.9
+import csv
+
+import boto3
+from mypy_boto3_s3.service_resource import Bucket
+
+
+@dataclass
+class VttsmaFileDownload:
+    """Use as CLI arguments for Dreem's library."""
+    # TODO: how does this relate to VTT?
+    directory: Path = config.storage_vol
+    # TODO: Check VTT's data file type and format and choose appopriately
+    ftype: str = "zip"
+
+
+# Define location and filetype to download
+args = VttsmaFileDownload()
+
+def get_bucket(creds: dict) -> Bucket:
+    """
+    Builds a S3 session bucket object to interface with the S3 bucket
+    """
+    session = boto3.session.Session(
+        aws_access_key_id=creds['aws_ak'], 
+        aws_secret_access_key=creds['aws_ask'],
+    )
+    
+    s3 = session.resource('s3')
+    bucket = s3.Bucket(creds['bucket_name'])    
+    
+    return bucket
+
+
+def get_list(bucket: Bucket) -> [dict]:
+    """
+    GET all records (metadata) from the AWS S3 bucket 
+
+    NOTE: S3 folder structure is symbolic. The 'key' (str) for each file object \
+        represents the path. See also `download_metadata()` in devices > vttsma.py 
+    """    
+    objects = bucket.objects.all()
+    object_paths = [obj.key for obj in objects]
+
+    # ignore users.txt files - data already present in object key
+    split_paths = [p.split('/') for p in object_paths if p.find('users.txt') == -1]
```

Comment:
Since we know the format, I wonder if it would be more readable to create a dataclass such that we can do:

```python
split_paths = [DMPRecord(**p.split('/')) for p in object_paths if p.find('users.txt') == -1] 
```

Also, instead of using find, might we use `not in p`, e.g., 

```python
[p.split('/') for p in object_paths if 'users.txt' not in p]
```

Reply:
I've messed around with this for too long - and think it overcomplicates it, especially for now. I'll add my current thoughts below and push these into a separate issue for future work.

### schemas > vttsma_record.py
```python

from datetime import datetime
from dataclasses import dataclass


@dataclass
class VTT_patient:
    """
    A patient-based object constructed from VTT S3 bucket metadata
    """
    vttsma_hash: str
    export_dates: [str]

    def __init__(self, s3_bucket_entry: str)
        """
        extracting values from s3_bucket_entry.split following
        dump_date, raw/files, patienthash, patienthash.nfo/.zip/.audio? 
        """
        values = s3_bucket_entry.split('/')
        self.vttsma_hash = values[2]
        self.export_dates = [values[0]]

    def __eq__(self, other):
        return (
            isinstance(other, self.__class__) and
            getattr(other, 'vttsma_hash', None) == self.vttsma_hash
        )
```

### lib > vttsma.py > get_list()

```python
records = [SMA_Record(r.key) for r in objects if 'users.txt' not in r.key]
# TODO: create a set which removes duplicates, but merges the dump_dates
```

Reply:
Tracking progress in Issue #26